Framework/MDAlgorithms/src/CreateMD.cpp

```diff
@@ -0,0 +1,468 @@
+      this->getProperty("DataSources");
+
+  const unsigned long entries = data_sources.size();
+
+  padParameterVector(psi, entries);
+  padParameterVector(gl, entries);
+  padParameterVector(gs, entries);
+  if (efix.empty()) {
+    efix.push_back(-1.0);
+  }
+  padParameterVector(efix, entries);
+
+  int counter = 0;
+  std::vector<std::string> to_merge_names;
+  std::string to_merge_name;
+  MatrixWorkspace_sptr workspace;
+  std::stringstream ws_name;
+  IMDEventWorkspace_sptr run_md;
+  for (unsigned long entry_number = 0; entry_number < entries;
+       ++entry_number, ++counter) {
+    ws_name.str(std::string());
+
+    // If data source is not an existing workspace it must be a file we need to
+    // load
+    if (!AnalysisDataService::Instance().doesExist(
+            data_sources[entry_number])) {
+      // Strip off any file extension or path to leave just the stem (base)
+      // filename
+      std::string filename_noext =
+          Poco::Path(data_sources[entry_number]).getBaseName();
+
+      // Create workspace name of form {filename}_md_{n}
+      ws_name << filename_noext << "_md_" << counter;
+      to_merge_name = ws_name.str();
+      workspace = boost::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(
+          loadWs(data_sources[entry_number], to_merge_name));
+    } else {
+      workspace = boost::dynamic_pointer_cast<Mantid::API::MatrixWorkspace>(
+          AnalysisDataService::Instance().retrieve(data_sources[entry_number]));
+      ws_name << data_sources[entry_number] << "_md";
+      to_merge_name = ws_name.str();
+    }
+
+    // We cannot process in place until we have an output MDWorkspace to use.
+    bool do_in_place = in_place && (counter > 0);
+    run_md = single_run(workspace, emode, efix[entry_number], psi[entry_number],
+                        gl[entry_number], gs[entry_number], do_in_place, alatt,
+                        angdeg, u, v, run_md);
+    to_merge_names.push_back(to_merge_name);
+
+    // We are stuck using ADS as we can't pass workspace pointers to MergeMD
+    // There is currently no way to pass a list of workspace pointers
+    AnalysisDataService::Instance().addOrReplace(to_merge_name, run_md);
+  }
+
+  Workspace_sptr output_workspace;
+  if (to_merge_names.size() > 1 && !in_place) {
+    output_workspace = merge_runs(to_merge_names);
+  } else {
+    output_workspace =
+        AnalysisDataService::Instance().retrieve(to_merge_names[0]);
+  }
+
+  // Clean up temporary workspaces
+  for (auto ws_iter = to_merge_names.begin(); ws_iter != to_merge_names.end();
+       ++ws_iter) {
+    AnalysisDataService::Instance().remove(*ws_iter);
+  }
+
+  this->setProperty("OutputWorkspace", output_workspace);
+}
+
+/*
+ * Load data from file
+ */
+Mantid::API::Workspace_sptr CreateMD::loadWs(const std::string &filename,
+                                             const std::string &wsname) {
+  Algorithm_sptr load_alg = createChildAlgorithm("Load");
+
```

Comment:
your methods are not changing any data members, so you can declare them `const`.